### PR TITLE
Add option to bypass authenticating to s3

### PR DESCRIPTION
**Pull request by jaustinpage**
This allows for anonymous uploads to s3.

An example would be a setup where you are running an s3 bucket that is accessed through an s3 endpoint on a vpn'ed subnet, or in some other highly restricted way.

Tested and "works on my machine", would love to see this added.

### CODE REVIEW

File: s3pypi/storage.py [hunk 1]
```diff
@@ -20,8 +22,16 @@ def __init__(
...
     ):
-        self.s3 = session.resource("s3", endpoint_url=s3_endpoint_url)
+        if no_sign_request:
+            self.s3 = session.resource(
+                "s3",
+                endpoint_url=s3_endpoint_url,
+                config=Config(signature_version=botocore.session.UNSIGNED)
+            )
+        else:
+            self.s3 = session.resource("s3", endpoint_url=s3_endpoint_url)
```
[the remark this answers is not in the document]
Comment (jaustinpage) [reply s3pypi/storage.py:34]: after running black, the formatting looks like this due to the line length:

```Python
        self.s3 = session.resource(
            "s3",
            endpoint_url=s3_endpoint_url,
            config=Config(signature_version=botocore.session.UNSIGNED)
            if no_sign_request
            else None,
        )
```

I can leave it like that, or i can do something like this:
```Python
        _config = None
        if no_sign_request:
            _config = Config(signature_version=botocore.session.UNSIGNED)

        self.s3 = session.resource(
            "s3",
            endpoint_url=s3_endpoint_url,
            config=_config,
        )
```
I think this avoids the duplicate `self.s3` assignment, while also not having an "inline conditional assignment" that is actually broken over multiple lines and looks kinda awkward...

Let me know either way, I am happy to do either one.

File: s3pypi/__main__.py [hunk 2]
```diff
@@ -62,6 +62,11 @@ def get_arg_parser():
         action="store_true",
         help="Write a root index that lists all available package names.",
     )
+    p.add_argument(
+        "--no-sign-request",
+        action="store_true",
+        help="Dont use authentication when communicating to s3.",
```
[the remark this answers is not in the document]
Comment (jaustinpage) [reply s3pypi/__main__.py:68]: I like it. I'll correct it.

File: s3pypi/storage.py [hunk 3]
```diff
@@ -4,6 +4,8 @@
 import boto3
 import botocore
 
+from botocore.config import Config
```
[the remark this answers is not in the document]
Comment (jaustinpage) [reply s3pypi/storage.py:7]: absolutely! i should have looked for a tox or makefile, mea culpa!

File: s3pypi/storage.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jaustinpage) [reply s3pypi/storage.py]: I went ahead and pushed (second option), if you would like me to switch it to (first option) I am happy to add another commit.

File: s3pypi/storage.py [hunk 4]
```diff
@@ -20,8 +21,17 @@ def __init__(
...
     ):
-        self.s3 = session.resource("s3", endpoint_url=s3_endpoint_url)
+        _config = None
+        if no_sign_request:
+            _config = Config(signature_version=botocore.session.UNSIGNED)
+
+        self.s3 = session.resource(
+            "s3",
+            endpoint_url=s3_endpoint_url,
+            config=_config,
```
Comment (mdwint) [inline s3pypi/storage.py:33]: This would fit on one line if you removed the trailing comma.

Comment (jaustinpage) [reply s3pypi/storage.py:33]: of course! Thanks, ill get that one fixed.

**Comment by jaustinpage** [comment]
If you want, i can squash commit 5ac254f and a824eeb into 1 commit if you want to make the commit history a little cleaner.

**Review by mdwint — APPROVED** [review]

**Comment by mdwint** [comment]
No need, @jaustinpage. Github will squash automatically on merge.

This looks good now. Thanks for your contribution!

**Pull request merged**